src/Quartz/Job/DirectoryScanJob.cs

```diff
@@ -57,95 +66,52 @@ public DirectoryScanJob()
-            IDirectoryScanListener listener = (IDirectoryScanListener) temp;
-
-            if (listener == null)
-            {
-                throw new JobExecutionException("DirectoryScanListener named '" +
-                                                listenerName + "' not found in SchedulerContext");
-            }
+            DirectoryScanJobModel model = DirectoryScanJobModel.GetInstance(context);
 
-            DateTime lastDate = DateTime.MinValue;
-            if (mergedJobDataMap.ContainsKey(LastModifiedTime))
+            List<FileInfo> updatedFiles = new List<FileInfo>();
+            Parallel.ForEach(model.DirectoriesToScan, d =>
             {
-                lastDate = mergedJobDataMap.GetDateTime(LastModifiedTime);
-            }
-
-            TimeSpan minAge = TimeSpan.FromSeconds(5);
-            if (mergedJobDataMap.ContainsKey(MinimumUpdateAge))
-            {
-                minAge = TimeSpan.FromMilliseconds(mergedJobDataMap.GetLong(MinimumUpdateAge));
-            }
-
-            DateTime maxAgeDate = DateTime.Now - minAge;
-
-            IEnumerable<FileInfo> updatedFiles = GetUpdatedOrNewFiles(dirName, lastDate, maxAgeDate);
-
-            if (updatedFiles == null)
-            {
-                log.Warn("Directory '" + dirName + "' does not exist.");
-            }
-
-            DateTime latestMod = lastDate;
-            foreach (FileInfo updFile in updatedFiles)
-            {
-                DateTime lm = updFile.LastWriteTime;
-                latestMod = (lm > latestMod) ? lm : latestMod;
-            }
+                var newFiles = GetUpdatedOrNewFiles(d, model.LastModTime, model.MaxAgeDate);
+                if (newFiles != null)
+                {
+                    updatedFiles.AddRange(GetUpdatedOrNewFiles(d, model.LastModTime, model.MaxAgeDate));
+                }
+            });
 
             if (updatedFiles.Any())
             {
                 // notify call back...
-                log.Info("Directory '" + dirName + "' contents updated, notifying listener.");
-                listener.FilesUpdatedOrAdded(updatedFiles);
+                updatedFiles.Select(x => x.DirectoryName)
+                    .Distinct().ToList()
+                    .ForEach(dir =>
+                    {
+                        log.Info("Directory '" + dir + "' contents updated, notifying listener.");
+                    });
+                model.DirectoryScanListener.FilesUpdatedOrAdded(updatedFiles);
+
+                DateTime latestWriteTimeFromFiles = updatedFiles.Select(x => x.LastWriteTime).Max();
+                model.UpdateLastModifiedDate(latestWriteTimeFromFiles);
             }
             else if (log.IsDebugEnabled())
             {
-                log.Debug("Directory '" + dirName + "' contents unchanged.");
+                model.DirectoriesToScan.ToList().ForEach(dir =>
+                {
+                    log.Debug("Directory '" + dir + "' contents unchanged.");
```

Comment:
The `DirectoryScanJobModel` currently does not have an `ILog` member, and it also is unaware of whether or not the directories have files in them or not. I thought of the model class as just being a way to get the data that the `DirectoryScanJob` needs in and out of the merged `JobDataMap`. The usage of that data (e.g. checking to see if the directories has files or not) is the job of the `DirectoryScanJob`. That is why I had the job logging if the directories did not contain any changed files, because it is aware of that fact. What do you think?